etc/gerbil.el

```diff
@@ -3,41 +3,55 @@
 (require 'scheme)
 (require 'cmuscheme)
 
-(defun gerbil-mode ()
+(defgroup gerbil-mode nil
+  "Editing Gerbil code"
+  :prefix "gerbil-mode-"
+  :group 'scheme)
+
+;; Redefine the function scheme-send-region from `cmuscheme' so
+;; that we can keep track of all text sent to Gambit's stdin.
+;; By Christopher Eames (Chream) <chream-gmx.com> 2018.
+
+(defun scheme-send-region (start end)
+  "Send the current region to the inferior Scheme process."
+  (interactive "r")
+  (scheme-send-string (buffer-substring start end)))
+
+(defun scheme-send-string (str)
+  "Send a string to the inferior Scheme process."
+  (gerbil-send-string str))
+
+(defun scheme-compile-region (start end)
   (interactive)
-  (kill-all-local-variables)
-  (gerbil-mode-init)
-  (scheme-mode-variables)
-  (run-hooks 'scheme-mode-hook)
-  (gerbil-init)
-  (run-hooks 'gerbil-mode-hook))
+  (gerbil-compile-current-buffer))
 
-(defun gerbil-mode-init ()
-  (use-local-map scheme-mode-map)
-  (setq major-mode 'gerbil-mode)
-  (setq mode-name "Gerbil"))
+(defun gerbil-message (string)
+  (message (concat "Gerbil-info : SENT=" string " ...")))
 
-(defgroup gerbil nil
-  "Editing Gerbil code"
-  :group 'scheme)
+(defun gerbil-send-string (string)
+  (let ((string (concat string "\n"))
+        (string-len (length string)))
+    (comint-check-source string)
+    (comint-send-string (scheme-proc) string)
+    (gerbil-message (subseq string 0 (string-match "\n" string)))))
 
-(defcustom gerbil-mode-hook nil
```

Comment:
are we losing the ability to run hooks?

Reply:
No, a mode hook is generated by the derived major mode definition.